File: leetcode/0098. Validate Binary Search Tree/Solution.swift
// https://leetcode.com/problems/validate-binary-search-tree/

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

class Solution {
    func isValidBST(_ root: TreeNode?, _ lowerBound: Int, _ upperBound: Int) -> Bool {
        guard let root = root else {
            return true
        }

        if root.val <= lowerBound || root.val >= upperBound {
            return false
        }

        return isValidBST(root.left, lowerBound, root.val) && isValidBST(root.right, root.val, upperBound)
    }

    func isValidBST(_ root: TreeNode?) -> Bool {
        return isValidBST(root, Int.min, Int.max)
    }
}

File: leetcode/0098. Validate Binary Search Tree/Solution.py
# https://leetcode.com/problems/validate-binary-search-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBSTInRange(self, root: TreeNode, a: float, b: float) -> bool:
        if not root:
            return True

        if root.val <= a or root.val >= b:
            return False

        return self.isValidBSTInRange(root.left, a, root.val) and self.isValidBSTInRange(root.right, root.val, b)

    def isValidBST(self, root: TreeNode) -> bool:
        return self.isValidBSTInRange(root, -float("inf"), float("inf"))

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def isNodeInRange(root: TreeNode, a: int, b: int) -> bool:
            if not root:
                return True
            
            if a >= root.val or root.val >= b:
                return False
            
            return isNodeInRange(root.left, a, root.val) and isNodeInRange(root.right, root.val, b)
            
        return isNodeInRange(root, -math.inf, math.inf)